--- a/build.sh
+++ b/build.sh
@@ -79,6 +79,18 @@ if [[ $CLEAN -eq 1 ]]; then
   echo ""
 fi
 
+# ── Build hygen backend ────────────────────────
+echo -e "${BOLD}  Building hygen backend...${RESET}"
+
+if [[ ! -d "hygen/build" ]]; then
+  run meson setup hygen/build hygen
+fi
+run ninja -C hygen/build
+if [[ $? -ne 0 ]]; then
+  fail "hygen build failed"
+  exit 1
+fi
+success "hygen backend built"
+echo ""
+
 # ── Build compiler ────────────────────────────
 echo -e "${BOLD}  Building compiler...${RESET}"
 
@@ -99,7 +111,11 @@ if [[ ! -d "compiler" ]]; then
     exit 1
   fi
 
-  run gcc lex.yy.c parser.tab.c ast.c ir.c lower.c opt.c codegen_asm.c  codegen_elf.c typecheck.c compiler.c -o ../hylian
+  run gcc lex.yy.c parser.tab.c ast.c ir.c ir_to_mir.c lower.c opt.c codegen_hygen.c typecheck.c compiler.c \
+    -I../hygen/libhymir/include -I../hygen/libhyregalloc/include -I../hygen/libhyx64/include -I../hygen/libhyobj/include \
+    ../hygen/build/libhymir/libhymir.a ../hygen/build/libhyregalloc/libhyregalloc.a \
+    ../hygen/build/libhyx64/libhyx64.a ../hygen/build/libhyobj/libhyobj.a \
+    -o ../hylian
 
   if [[ $? -ne 0 ]]; then
     fail "gcc compilation failed"
